Drop keyword support: small fixes

gwgundersen opened the pull request:
- Addresses issue https://github.com/pydata/xarray/issues/2910 and completes PR https://github.com/pydata/xarray/pull/3128.
 - Passes `black .` and `flake8`.
 - Documented in `whats-new.rst` and `indexing.rst`.

gwgundersen commented:
Strange. This passes all my tests locally.

max-sixty commented on doc/indexing.rst:
```diff
@@ -238,7 +238,7 @@ index labels along a dimension dropped:
 .. ipython:: python
     :okwarning:
```
We can remove this now

max-sixty commented:
The failure is b/c of an upstream package; we can ignore

shoyer commented on xarray/core/dataset.py:
```diff
@@ -3558,6 +3557,13 @@ def drop(  # noqa: F811
                 ds = ds._drop_labels(labels, dim, errors=errors)
             return ds
         elif dim is None:
+            if isinstance(labels, DataArrayCoordinates):
```
I think this should probably be even more generic:
```suggestion
            if is_dict_like(labels) and not isinstance(labels, dict):
```
and probably moved up to ensure it gets triggered in all situations, not just if `dim is None`.

gwgundersen replied:
Sure, makes sense. I tried to write an error message that captured the generic case as well.

shoyer commented on xarray/core/dataset.py:
```diff
@@ -3558,6 +3557,13 @@ def drop(  # noqa: F811
+                warnings.warn(
+                    "dropping coordinates using a DataArrayCoordinates object "
+                    "is deprecated; use drop_vars or a list of coordinates",
+                    DeprecationWarning,
```
`FutureWarning` is more appropriate here, since we will probably break existing code in the future by giving this different behavior:
```suggestion
                    FutureWarning,
```

shoyer replied:
It also would be great to add a test that verifies that the warning is raised :)

gwgundersen replied:
Should the second warning for `list`-like `labels` also be a `FutureWarning`?

shoyer replied:
I think `DeprecationWarning` might be more appropriate there, since continuing to rely on that will result in errors rather than different behavior.

gwgundersen commented:
@shoyer, I've addressed your comments by adding more helpful warnings and testing that said warnings are raised properly. I've also moved both warnings to the top of the function as I think it's a bit easier to reason about the code branches this way.

pep8speaks commented:
Hello @gwgundersen! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`xarray/core/dataset.py`](https://github.com/pydata/xarray/blob/34f226709afd34fcff9845002daabc6cf6fec50f/xarray/core/dataset.py):

> [Line 57:1](https://github.com/pydata/xarray/blob/34f226709afd34fcff9845002daabc6cf6fec50f/xarray/core/dataset.py#L57): [F401](https://duckduckgo.com/?q=pep8%20F401) '.coordinates.DataArrayCoordinates' imported but unused
